sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -297,7 +297,7 @@ object InMemoryFileIndex extends Logging {
     val missingFiles = mutable.ArrayBuffer.empty[String]
     val filteredLeafStatuses = allLeafStatuses.filterNot(
       status => shouldFilterOut(status.getPath.getName))
-    val resolvedLeafStatuses = filteredLeafStatuses.flatMap {
+    val resolvedLeafStatuses = filteredLeafStatuses.par.flatMap {
```

Comment:
As this can be called on executors, I think we should use `ThreadUtils.parmap`. cc @MaxGekk

Reply:
btw, is this a right approach? I a little confuse this with the current parallel partition discovery path..

Reply:
Thanks @viirya for feedback. Yes, this method can be called on executors as below. Do you think it's not thread-safe ?
Each partitions will have its own hadoopConf and then own fs, and nothing is shared in this method.

sparkContext
        .parallelize(serializedPaths, numParallelism)
        .mapPartitions { pathStrings =>
          val hadoopConf = serializableConfiguration.value
          pathStrings.map(new Path(_)).toSeq.map { path =>
            (path, listLeafFiles(path, hadoopConf, filter, None))
          }.iterator
        }.map { case (path, statuses) =>

Reply:
Thanks @maropu for your comments. I updated the title and description. Let's explain the difference between this change and the current parallel partition discovery. The current one will discovery different partitions in parallel. This change will get the block location for a single partition in parallel. When there is only a few partitions and each contains tons of thousands of files, the current partition discovery won't help. And this change can accelerate it in this case

Reply:
Parallel Scala collections are not interruptible in some cases as a consequence of that if you use them on executors, tasks cannot be canceled properly. You can do an experiment yourself and run the code in a lambda function: https://github.com/apache/spark/blob/131ca146ed390cd0109cd6e8c95b61e418507080/core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala#L143-L150

When you cancel the job, threads will be still blocking on the sleep call.